lighthouse-core/audits/script-treemap-data.js

```diff
@@ -0,0 +1,262 @@
+   */
+  static get meta() {
+    return {
+      id: 'script-treemap-data',
+      scoreDisplayMode: Audit.SCORING_MODES.INFORMATIVE,
+      title: 'Script Treemap Data',
+      description: 'Used for treemap app',
+      requiredArtifacts:
+        ['traces', 'devtoolsLogs', 'SourceMaps', 'ScriptElements', 'JsUsage', 'URL'],
+    };
+  }
+
+  /**
+   * Returns a tree data structure where leaf nodes are sources (ie. real files from source tree)
+   * from a source map, and non-leaf nodes are directories. Leaf nodes have data
+   * for bytes, coverage, etc., when available, and non-leaf nodes have the
+   * same data as the sum of all descendant leaf nodes.
+   * @param {string} sourceRoot
+   * @param {Record<string, SourceData>} sourcesData
+   * @return {Node}
+   */
+  static prepareTreemapNodes(sourceRoot, sourcesData) {
+    /**
+     * @param {string} name
+     * @return {Node}
+     */
+    function newNode(name) {
+      return {
+        name,
+        resourceBytes: 0,
+      };
+    }
+
+    const topNode = newNode(sourceRoot);
+
+    /**
+     * Given a slash-delimited path, traverse the Node structure and increment
+     * the data provided for each node in the chain. Creates nodes as needed.
+     * Ex: path/to/file.js will find or create "path" on `node`, increment the data fields,
+     *     and continue with "to", and so on.
+     * @param {string} source
+     * @param {SourceData} data
+     */
+    function addAllNodesInSourcePath(source, data) {
+      let node = topNode;
+
+      // Apply the data to the topNode.
+      topNode.resourceBytes += data.resourceBytes;
+      if (data.unusedBytes) topNode.unusedBytes = (topNode.unusedBytes || 0) + data.unusedBytes;
+
+      // Strip off the shared root.
+      const sourcePathSegments = source.replace(sourceRoot, '').split(/\/+/);
+      sourcePathSegments.forEach((sourcePathSegment, i) => {
+        const isLastSegment = i === sourcePathSegments.length - 1;
+
+        let child = node.children && node.children.find(child => child.name === sourcePathSegment);
+        if (!child) {
+          child = newNode(sourcePathSegment);
+          node.children = node.children || [];
+          node.children.push(child);
+        }
+        node = child;
+
+        // Now that we've found or created the next node in the path, apply the data.
+        node.resourceBytes += data.resourceBytes;
+        if (data.unusedBytes) node.unusedBytes = (node.unusedBytes || 0) + data.unusedBytes;
+
+        // Only leaf nodes might have duplication data.
+        if (isLastSegment && data.duplicatedNormalizedModuleName !== undefined) {
+          node.duplicatedNormalizedModuleName = data.duplicatedNormalizedModuleName;
+        }
+      });
+    }
+
+    // For every source file, apply the data to all components
+    // of the source path, creating nodes as necessary.
+    for (const [source, data] of Object.entries(sourcesData)) {
+      addAllNodesInSourcePath(source || `<unmapped>`, data);
```

Comment:
is the `<unmapped>` fallback possible? It seems like things would be pretty broken if the source name was the empty string

Reply:
ah, nope. not possible.

There is an entry in `sourcesWastedBytes` for `(unmapped)` (see https://github.com/GoogleChrome/lighthouse/blob/8dd649aa30553b6862229fcee57d8c2039500b70/lighthouse-core/computed/unused-javascript-summary.js#L135 / https://github.com/GoogleChrome/lighthouse/blob/7008e5502ec4465d47f15d54fa25383a4b0db40a/lighthouse-core/test/audits/byte-efficiency/unused-javascript-test.js#L123 ), but atm this audit doesn't surface unmapped bytes. The source datas are made by iterating the bundle files, which will always have a non empty string.

will add an item in parent issue to add a node for unmapped bytes.